File: server/storage.ts
import { sessions, users, type Session, type InsertSession, type User, type InsertUser } from "@shared/schema";
import { db } from "./db";
import { eq, desc } from "drizzle-orm";
import bcrypt from "bcryptjs";

export interface IStorage {
  createSession(topic: string, userId?: string): Promise<Session>;
  getSession(id: string): Promise<Session | undefined>;
  getAllSessions(userId?: string): Promise<Session[]>;
  getUserSessions(userId: string): Promise<Session[]>;
  updateSession(id: string, data: any): Promise<Session>;
  createUser(email: string, password: string, name?: string): Promise<User>;
  getUser(email: string): Promise<User | undefined>;
  getUserById(id: string): Promise<User | undefined>;
}

export class DatabaseStorage implements IStorage {
  async createSession(topic: string, userId?: string): Promise<Session> {
    const [session] = await db
      .insert(sessions)
      .values({
        topic,
        userId,
        duration: 0,
        eyeContactPercentage: 0,
        confidenceScore: 0,
        wordsPerMinute: 0,
        fillerWordsCount: 0,
        transcript: '',
        strengths: [],
        improvements: [],
        eyeContactData: [],
      })
      .returning();
    return session;
  }

  async getSession(id: string): Promise<Session | undefined> {
    const [session] = await db.select().from(sessions).where(eq(sessions.id, id));
    return session || undefined;
  }

  async getAllSessions(userId?: string): Promise<Session[]> {
    if (userId) {
      return this.getUserSessions(userId);
    }
    const allSessions = await db.select().from(sessions).orderBy(desc(sessions.createdAt));
    return allSessions;
  }

  async getUserSessions(userId: string): Promise<Session[]> {
    const userSessions = await db
      .select()
      .from(sessions)
      .where(eq(sessions.userId, userId))
      .orderBy(desc(sessions.createdAt));
    return userSessions;
  }

  async updateSession(id: string, data: any): Promise<Session> {
    const [session] = await db
      .update(sessions)
      .set(data)
      .where(eq(sessions.id, id))
      .returning();
    return session;
  }

  async createUser(email: string, password: string, name?: string): Promise<User> {
    const hashedPassword = await bcrypt.hash(password, 10);
    const [user] = await db
      .insert(users)
      .values({
        email,
        password: hashedPassword,
        name,
      })
      .returning();
    return user;
  }

  async getUser(email: string): Promise<User | undefined> {
    const [user] = await db.select().from(users).where(eq(users.email, email));
    return user || undefined;
  }

  async getUserById(id: string): Promise<User | undefined> {
    const [user] = await db.select().from(users).where(eq(users.id, id));
    return user || undefined;
  }
}

export const storage = new DatabaseStorage();

File: server-fastapi/storage.py
# server-fastapi/storage.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update
from sqlalchemy.sql import desc
from typing import Optional, List, Dict, Any
import uuid

from models import User, Session
from auth import hash_password

class DatabaseStorage:
    """
    Database storage operations
    Matches: DatabaseStorage class from storage.ts
    """
    
    async def create_session(
        self, 
        topic: str, 
        user_id: Optional[str],
        db: AsyncSession
    ) -> Session:
        """Create new practice session"""
        new_session = Session(
            id=str(uuid.uuid4()),
            topic=topic,
            user_id=user_id,
            duration=0,
            eye_contact_percentage=0,
            confidence_score=0,
            words_per_minute=0,
            filler_words_count=0,
            transcript='',
            strengths=[],
            improvements=[],
            eye_contact_data=[],
        )
        
        db.add(new_session)
        await db.commit()
        await db.refresh(new_session)
        return new_session
    
    async def get_session(self, session_id: str, db: AsyncSession) -> Optional[Session]:
        """Get session by ID"""
        result = await db.execute(
            select(Session).where(Session.id == session_id)
        )
        return result.scalar_one_or_none()
    
    async def get_all_sessions(
        self, 
        user_id: Optional[str],
        db: AsyncSession
    ) -> List[Session]:
        """Get all sessions, optionally filtered by user"""
        if user_id:
            return await self.get_user_sessions(user_id, db)
        
        result = await db.execute(
            select(Session).order_by(desc(Session.created_at))
        )
        return result.scalars().all()
    
    async def get_user_sessions(self, user_id: str, db: AsyncSession) -> List[Session]:
        """Get all sessions for a specific user"""
        result = await db.execute(
            select(Session)
            .where(Session.user_id == user_id)
            .order_by(desc(Session.created_at))
        )
        return result.scalars().all()
    
    async def update_session(
        self, 
        session_id: str, 
        data: Dict[str, Any],
        db: AsyncSession
    ) -> Session:
        """Update session with analysis results"""
        await db.execute(
            update(Session)
            .where(Session.id == session_id)
            .values(**data)
        )
        await db.commit()
        
        # Fetch and return updated session
        return await self.get_session(session_id, db)
    
    async def create_user(
        self, 
        email: str, 
        password: str, 
        name: Optional[str],
        db: AsyncSession
    ) -> User:
        """Create new user with hashed password"""
        hashed_password = hash_password(password)
        
        new_user = User(
            id=str(uuid.uuid4()),
            email=email,
            password=hashed_password,
            name=name
        )
        
        db.add(new_user)
        await db.commit()
        await db.refresh(new_user)
        return new_user
    
    async def get_user(self, email: str, db: AsyncSession) -> Optional[User]:
        """Get user by email"""
        result = await db.execute(
            select(User).where(User.email == email)
        )
        return result.scalar_one_or_none()
    
    async def get_user_by_id(self, user_id: str, db: AsyncSession) -> Optional[User]:
        """Get user by ID"""
        result = await db.execute(
            select(User).where(User.id == user_id)
        )
        return result.scalar_one_or_none()

# Global storage instance
storage = DatabaseStorage()
